src/python/pants/backend/python/lint/isort/rules.py

```diff
@@ -36,11 +51,74 @@ def opt_out(cls, tgt: Target) -> bool:
 class IsortRequest(FmtTargetsRequest):
     field_set_type = IsortFieldSet
     tool_subsystem = Isort
-    partitioner_type = PartitionerType.DEFAULT_SINGLE_PARTITION
+    partitioner_type = PartitionerType.CUSTOM
+
+
+@dataclass(frozen=True)
+class IsortPartitionMetadata:
+    path_to_firstparty_modules: FrozenDict[str, tuple[str, ...]]
+
+    @property
+    def description(self) -> str | None:
+        return None
+
+
+@rule(desc="Partition isort", level=LogLevel.DEBUG)
+async def isort_partition(
+    request: IsortRequest.PartitionRequest, isort: Isort
+) -> Partitions[str, IsortPartitionMetadata]:
+    if isort.skip:
+        return Partitions()
+
+    all_sources_paths = await MultiGet(
+        Get(SourcesPaths, SourcesPathsRequest(field_set.source)) for field_set in request.field_sets
+    )
+
+    metadata = defaultdict(set)
+    if isort.firstparty_discovery:
+        direct_dependencies = await MultiGet(
+            Get(Targets, DependenciesRequest(field_set.dependencies))
+            for field_set in request.field_sets
+        )
+        all_direct_targets = {*itertools.chain.from_iterable(direct_dependencies)}
+        stripped_files = await MultiGet(
+            Get(
+                StrippedFileName,
+                StrippedFileNameRequest(
+                    target.get(PythonSourceField, default_raw_value="").file_path
```

Comment:
Rather than default to `""` and run the request, maybe we could this filter out targets that don't have `PythonSourceField` (e.g. resources, `python_requirements` etc.) from `all_direct_targets`?